In [1]:
import pandas as pd

In [10]:
# Load the CSV file into a DataFrame
patient_21 = "21_Patient_backup(in).csv"
patient = pd.read_csv(patient_21)
patient.head()


id  from_pms             sikka_sub_task sikka_sub_task_operation  \
0  7732      True  SikkaTaskLog object (922)                   update   
1  8221      True  SikkaTaskLog object (922)                   update   
2  3042      True  SikkaTaskLog object (922)                   update   
3  3685      True  SikkaTaskLog object (922)                   update   
4  4120      True  SikkaTaskLog object (922)                   update   

   last_updated  user                                  uuid  patient_id  \
0           NaN   NaN  f4702dc1-9a38-4a7d-81df-c3e8f4cb2583      5801.0   
1           NaN   NaN  cb55c5c0-ca46-41a3-b2a6-7d53d124c9e0      5312.0   
2           NaN   NaN  d9b4cf2f-4a17-4cc8-977a-28bf86d8f227     10493.0   
3           NaN   NaN  abdc31fb-9972-4e35-afb6-2282b282d41f      9850.0   
4           NaN   NaN  240e5738-02f2-40a3-bd55-b523c131a3d2      9415.0   

          guarantor                              office  ... practice  \
0  ANTONIO PELACCIA  21-Dentalhouse - Woodbridge Dental  ...      1.0   
1     LOUISE MIAZGA  21-Dentalhouse - Woodbridge Dental  ...      1.0   
2      KATHY HUFSKY  21-Dentalhouse - Woodbridge Dental  ...      1.0   
3       KEVIN COOZE  21-Dentalhouse - Woodbridge Dental  ...      1.0   
4   MICHAEL PALUMBO  21-Dentalhouse - Woodbridge Dental  ...      1.0   

  sikka_provider_id primary_insurance_company_id insurance balance cust_id  \
0                 D                          NaN       NaN     NaN     NaN   
1                 D                          NaN       NaN     NaN     NaN   
2               NaN                       HUSKY1       NaN     NaN     NaN   
3                KG                       LITENS       NaN     NaN     NaN   
4                KG                     GANZBROT       NaN     NaN     NaN   

  gender provider                      profile_picture  \
0    NaN      NaN  profile_pictures/sample_profile.png   
1    NaN      NaN  profile_pictures/sample_profile.png   
2    NaN      NaN  profile_pictures/sample_profile.png   
3    NaN      NaN  profile_pictures/sample_profile.png   
4    NaN      NaN  profile_pictures/sample_profile.png   

                        date_joined  
0  2024-07-24 12:15:18.995078+00:00  
1  2024-07-24 12:15:18.995078+00:00  
2  2024-07-24 12:15:18.995078+00:00  
3  2024-07-24 12:15:18.995078+00:00  
4  2024-07-24 12:15:18.995078+00:00  

[5 rows x 35 columns]

In [11]:
guarantor_21 = "21_Guarantor_backup(in).csv"
guarantor = pd.read_csv(guarantor_21)
guarantor.head()

id  from_pms             sikka_sub_task sikka_sub_task_operation  \
0   6256      True  SikkaTaskLog object (919)                   update   
1   6110      True  SikkaTaskLog object (919)                   update   
2  21617      True  SikkaTaskLog object (919)                   create   
3   5760      True  SikkaTaskLog object (919)                   update   
4   6492      True  SikkaTaskLog object (919)                   update   

   last_updated  guarantor_id  user                              office  \
0           NaN        4032.0   NaN  21-Dentalhouse - Woodbridge Dental   
1           NaN        4292.0   NaN  21-Dentalhouse - Woodbridge Dental   
2           NaN       13680.0   NaN  21-Dentalhouse - Woodbridge Dental   
3           NaN        4879.0   NaN  21-Dentalhouse - Woodbridge Dental   
4           NaN        3652.0   NaN  21-Dentalhouse - Woodbridge Dental   

  firstname middlename  ...    status         address_line1  \
0      REMO        NaN  ...  Inactive    144 Vineyard Court   
1      JUDY        NaN  ...  Inactive                   NaN   
2    SOPHIA        NaN  ...    Active                   NaN   
3     LAURA        NaN  ...  Inactive                   NaN   
4     FRANK        NaN  ...  Inactive  47 Venetian Crescent   

                 city state zipcode          cell                    email  \
0  Woodbridge Ontario    ON  L4L4M6           NaN                      NaN   
1   Thornhill Ontario    ON  L4J1K3           NaN                      NaN   
2                 NaN   NaN     NaN  6.472803e+09  sophia.salini@gmail.com   
3                 NaN   NaN     NaN           NaN                      NaN   
4  Woodbridge Ontario    ON  L4L5E6           NaN                      NaN   

  employer_id  guarantor_s_patient_id                      profile_picture  
0         0.0                     NaN  profile_pictures/sample_profile.png  
1         0.0                     NaN  profile_pictures/sample_profile.png  
2         0.0                     NaN  profile_pictures/sample_profile.png  
3         0.0                     NaN  profile_pictures/sample_profile.png  
4         0.0                     NaN  profile_pictures/sample_profile.png  

[5 rows x 23 columns]

In [7]:
patient.columns

Index(['id', 'from_pms', 'sikka_sub_task', 'sikka_sub_task_operation',
       'last_updated', 'user', 'uuid', 'patient_id', 'guarantor', 'office',
       'fullname', 'first_name', 'middle_name', 'last_name', 'birth_date',
       'address', 'city', 'state', 'country', 'first_visit', 'last_visit',
       'last_login', 'email', 'phone', 'zip_code', 'practice',
       'sikka_provider_id', 'primary_insurance_company_id', 'insurance',
       'balance', 'cust_id', 'gender', 'provider', 'profile_picture',
       'date_joined'],
      dtype='object')

In [12]:
guarantor.columns

Index(['id', 'from_pms', 'sikka_sub_task', 'sikka_sub_task_operation',
       'last_updated', 'guarantor_id', 'user', 'office', 'firstname',
       'middlename', 'lastname', 'salutation', 'birthdate', 'status',
       'address_line1', 'city', 'state', 'zipcode', 'cell', 'email',
       'employer_id', 'guarantor_s_patient_id', 'profile_picture'],
      dtype='object')

## Can the fullname of guarantor be a matching key?

In [28]:
# check for basic information of id-type columns
def validate_name_column(data, column_name):
    # Check for missing values
    missing_values = data[column_name].isnull().sum()
    
    # Check for empty strings or invalid entries (non-numeric values, if applicable)
    invalid_values = data[column_name].apply(lambda x: isinstance(x, str) and x.strip() == "").sum()
    
    # Check for duplicate IDs
    duplicate_ids = data[column_name].duplicated().sum()
    
    # Print the results
    print(f"# of missing name: {missing_values}")
    print(f"Invalid values: {invalid_values}")
    print(f"Duplicate values: {duplicate_ids}")

In [30]:
validate_name_column(patient, 'guarantor')

# of missing name: 162
Invalid values: 112
Duplicate values: 5825


In [31]:
validate_name_column(guarantor, 'firstname')

# of missing name: 320
Invalid values: 0
Duplicate values: 5652


In [33]:
validate_name_column(guarantor, 'lastname')

# of missing name: 115
Invalid values: 0
Duplicate values: 3536


## Try to use guarantor name as a matching key

In [38]:
# Create a copy of the guarantor dataset
data_guarantor = guarantor.copy()

data_guarantor['guarantor_fullname'] = (
    data_guarantor['firstname'].fillna('') + ' ' +
    data_guarantor['middlename'].fillna('') + ' ' +
    data_guarantor['lastname'].fillna('')
)

# Remove any extra spaces if 'middlename' is missing
data_guarantor['guarantor_fullname'] = data_guarantor['guarantor_fullname'].str.strip()

data_guarantor[['firstname', 'middlename', 'lastname', 'guarantor_fullname']].head()



firstname middlename     lastname guarantor_fullname
0      REMO        NaN      CAMPOLI      REMO  CAMPOLI
1      JUDY        NaN  DEPELLEGRIN  JUDY  DEPELLEGRIN
2    SOPHIA        NaN       SALINI     SOPHIA  SALINI
3     LAURA        NaN      GIRALDI     LAURA  GIRALDI
4     FRANK        NaN     IMBROGNO    FRANK  IMBROGNO

In [43]:
# Check if the name could be matched

# Standardize the text in both datasets (convert to lowercase and strip spaces)
patient['guarantor_clean'] = patient['guarantor'].str.lower().str.strip()
data_guarantor['guarantor_fullname_clean'] = data_guarantor['guarantor_fullname'].str.lower().str.strip()

# Perform the match
matches = patient['guarantor_clean'].isin(data_guarantor['guarantor_fullname_clean'])

# Add a column to the patient dataset indicating match status
patient['Match_Status'] = matches

# Check the results
print(f"Total rows in patient dataset: {len(patient)}")
print(f"Number of matches found: {matches.sum()}")
print(f"Number of unmatched rows: {len(patient) - matches.sum()}")

Total rows in patient dataset: 13665
Number of matches found: 315
Number of unmatched rows: 13350


In [45]:
# Merge the data if the name could match (for test)
data_patient = patient.copy()
merged_data = pd.merge(data_patient, data_guarantor, 
    left_on='guarantor_clean', 
    right_on='guarantor_fullname_clean', 
    how='inner',  # Use 'inner' to keep only matched rows
    suffixes=('_patient', '_guarantor')
)


print("Merged dataset:")
merged_data.head()

Merged dataset:


id_patient  from_pms_patient     sikka_sub_task_patient  \
0        7744              True  SikkaTaskLog object (922)   
1        4849              True  SikkaTaskLog object (922)   
2        5259              True  SikkaTaskLog object (922)   
3        5259              True  SikkaTaskLog object (922)   
4        5259              True  SikkaTaskLog object (922)   

  sikka_sub_task_operation_patient  last_updated_patient  user_patient  \
0                           update                   NaN           NaN   
1                           update                   NaN           NaN   
2                           update                   NaN           NaN   
3                           update                   NaN           NaN   
4                           update                   NaN           NaN   

                                   uuid  patient_id       guarantor  \
0  9d72a7f2-5be1-4573-a170-a07d77975f45      5789.0   BASEBALL TEAM   
1  949af014-77d3-4c0e-80a5-a699a060e9c1      8686.0            CIAR   
2  857d8f3b-0544-4614-a562-1f221cce7729      8274.0                   
3  857d8f3b-0544-4614-a562-1f221cce7729      8274.0                   
4  857d8f3b-0544-4614-a562-1f221cce7729      8274.0                   

                       office_patient  ... city_guarantor state_guarantor  \
0  21-Dentalhouse - Woodbridge Dental  ...            NaN             NaN   
1  21-Dentalhouse - Woodbridge Dental  ...            NaN             NaN   
2  21-Dentalhouse - Woodbridge Dental  ...            NaN             NaN   
3  21-Dentalhouse - Woodbridge Dental  ...            NaN             NaN   
4  21-Dentalhouse - Woodbridge Dental  ...            NaN             NaN   

  zipcode cell email_guarantor employer_id guarantor_s_patient_id  \
0     NaN  NaN             NaN         0.0                    NaN   
1     NaN  NaN             NaN         0.0                    NaN   
2     NaN  NaN             NaN         0.0                    NaN   
3     NaN  NaN             NaN         0.0                    NaN   
4     NaN  NaN             NaN         0.0                    NaN   

             profile_picture_guarantor  guarantor_fullname  \
0  profile_pictures/sample_profile.png       BASEBALL TEAM   
1  profile_pictures/sample_profile.png                CIAR   
2  profile_pictures/sample_profile.png                       
3  profile_pictures/sample_profile.png                       
4  profile_pictures/sample_profile.png                       

  guarantor_fullname_clean  
0            baseball team  
1                     ciar  
2                           
3                           
4                           

[5 rows x 62 columns]

The column "guarantor_s_patient_id" is empty, could be a potential variable.

## Validation check through Patient data Birth_date 

In [78]:
from datetime import datetime
today = datetime.today()

patient['birth_date'] = pd.to_datetime(patient['birth_date'], errors='coerce')
patient['age'] = patient['birth_date'].apply(lambda x: today.year - x.year - ((today.month, today.day) < (x.month, x.day)) if pd.notnull(x) else None)
filtered_patients = patient[(patient['age'] <= 3) | (patient['age'] >= 80)]

print("Filtered dataset with valid ages (3 to 80):")
print(filtered_patients[['id', 'fullname', 'birth_date', 'age']])

Filtered dataset with valid ages (3 to 80):
          id            fullname                birth_date    age
0       7732      MARIA PELACCIA 1934-12-09 00:00:00+00:00   90.0
6       8409    EMILIO GIAMMARIA 1940-11-06 00:00:00+00:00   84.0
16     10059       ANTHONY FALBO 1944-03-09 00:00:00+00:00   80.0
26      7668      CARMELA ROMOLO 1927-07-11 00:00:00+00:00   97.0
29      9041      FRANK CERISANO 1925-04-25 00:00:00+00:00   99.0
...      ...                 ...                       ...    ...
13655  10311  GIOVANNI IACOBELLI 1924-10-01 00:00:00+00:00  100.0
13656  10312      BRUNA GREGORIS 1939-04-22 00:00:00+00:00   85.0
13660  12796          JOHN MASCI 1927-08-21 00:00:00+00:00   97.0
13663  13103     VIVI GIALEDAKIS 1940-11-21 00:00:00+00:00   84.0
13664  13104  EMANUEL GIALEDAKIS 1937-12-27 00:00:00+00:00   87.0

[1856 rows x 4 columns]


1856 rows are potentially invalid, they are all group of "too old"; could cosider to remove them later

## Insuerance ID validation Check

In [79]:
missing_insurance_count = patient['primary_insurance_company_id'].isnull().sum()
print(f"Number of missing values in 'primary_insurance_company_id': {missing_insurance_count}")

Number of missing values in 'primary_insurance_company_id': 5364


In [83]:
unique_insurance_ids = patient['primary_insurance_company_id'].dropna().unique()

array(['HUSKY1', 'LITENS', 'GANZBROT', ..., 'BENNE', 'PANSON1', 'BROCK 2'],
      dtype=object)